src/lambdas/auditMessageFirehoseReingest/handler.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  Context,
+  SQSBatchItemFailure,
+  SQSBatchResponse,
+  SQSEvent,
+  SQSRecord
+} from 'aws-lambda'
+import { initialiseLogger, logger } from '../../services/logger'
+import { S3ObjectDetails } from '../../types/s3ObjectDetails'
+import { tryParseJSON } from '../../utils/helpers/tryParseJson'
+import { deleteOrUpdateS3Objects } from './deleteOrUpdateS3Objects'
+import { getAuditEvents } from './getAuditEvents'
+import { sendAuditEventsToFirehose } from './sendAuditEventsToFirehose'
+
+export const handler = async (
+  event: SQSEvent,
+  context: Context
+): Promise<SQSBatchResponse> => {
+  initialiseLogger(context)
+
+  logger.info(`Received ${event.Records.length} records for processing`)
+
+  const batchItemFailures: SQSBatchItemFailure[] = []
+  const s3ObjectDetails: S3ObjectDetails[] = getS3ObjectDetails(event.Records)
+
+  // Get the audit events from the S3 objects, if there is a failure we can send
+  // the message back to the queue for reprocessing
+  const getAuditEventsResults = await getAuditEvents(s3ObjectDetails)
+  batchItemFailures.push(
+    ...messageIdsToBatchItemFailures(getAuditEventsResults.failedIds)
+  )
+
+  if (batchItemFailures.length > 0) {
+    logger.warn(
+      'Finished retrieving Audit Events from S3. There were some failures.',
+      { batchItemFailures }
+    )
+  } else {
+    logger.info('Successfully retrieved Audit Events from S3')
+  }
+
+  // Send the audit events to Firehose, and collect any that Firehose failed
+  // to ingest
+  const sendAuditEventsToFirehoseResults = await sendAuditEventsToFirehose(
+    getAuditEventsResults.successfulResults
+  )
+
+  // Delete or update the S3 objects based on the results of the Firehose
+  // ingestion. If there are any failures we can keep the S3 object and update
+  // it with the events that failed to reingest
+  await deleteOrUpdateS3Objects(sendAuditEventsToFirehoseResults)
+
+  return {
+    batchItemFailures
+  }
+}
+
+const getS3ObjectDetails = (records: SQSRecord[]): S3ObjectDetails[] => {
+  const s3ObjectDetails = records
+    .filter((record) => {
+      const isS3TestEvent = tryParseJSON(record.body).Event === 's3:TestEvent'
```

Comment:
The logic for working out if this is a test event could go in a function to bring down the line count of this function a bit

Reply:
Ok filters moved. How is that?